File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1531-string-compression-ii/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1531-string-compression-ii.py
class Solution:
    def getLengthOfOptimalCompression(self, s: str, k: int) -> int:
        
        def cost(streak):
            if streak == 0:
                return 0
            if streak == 1:
                return 1
            if streak <= 9:
                return 2
            if streak <= 99:
                return 3
            return 4
        
        @cache
        def dfs(index, pchar, delete, streak):
            if index >= len(s):
                return cost(streak)
            
            mincost = inf
            char = s[index]
            
            # Take
            if char == pchar:
                mincost = min(mincost, dfs(index+1, char, delete, streak + 1))
            else:
                mincost = min(mincost, cost(streak) + dfs(index+1, char, delete, 1))
            
            # Dont take
            if delete > 0:
                mincost = min(mincost, dfs(index+1, pchar, delete - 1, streak))
                
            return mincost
        
        return dfs(0, '#', k, 0)

File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1531-string-compression-ii/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1531-string-compression-ii.java
class Solution {
    String s;
    Map<String,Integer> map;
    int n;
    public int getLengthOfOptimalCompression(String s, int k) {
        this.s = s;
        this.n = s.length();
        map = new HashMap<>();
        return dfs(0,'#',0,k);
    }
    
    int dfs(int idx, char pChar, int streak, int delete){
        if(idx >= n){
            return cost(streak);
        }
        
        String key = idx+"-"+pChar+"-"+streak+"-"+delete;
        if(map.containsKey(key)) return map.get(key);
        
        char c = s.charAt(idx);
        
        int minCount = Integer.MAX_VALUE;
        
        //take
        if(c==pChar){
            minCount = Math.min(minCount,dfs(idx+1,c,streak+1,delete));
        } else{
            minCount = Math.min(minCount,cost(streak) + dfs(idx+1,c,1,delete));
        }
        
        //dont
        if(delete>=1){
            minCount = Math.min(minCount,dfs(idx+1,pChar,streak,delete-1));
        }
        
        map.put(key,minCount);
        
        return minCount;
    }
    
    int cost(int streak){
        if(streak==0) return 0;
        else if(streak==1) return 1;
        else if(streak<=9) return 2;
        else if(streak<=99) return 3;
        else return 4;
    }
        
}